### PR TITLE
Launch from build directory fix, and can scroll output

### PR DESCRIPTION
**Pull request by rohit-kumar-j**
Solves https://github.com/Civitasv/cmake-tools.nvim/pull/47#issuecomment-1545068911

The intended behaviour for #34(on by default), but also keeps the old behaviour :D

### CODE REVIEW

File: lua/cmake-tools/utils.lua [hunk 1]
```diff
@@ -68,16 +68,41 @@ function utils.execute(executable, opts)
...
-  local prefix = string.format("%s %d new", opts.cmake_console_position, opts.cmake_console_size)
-
-  utils.close_cmake_console();
-  vim.cmd(prefix .. " | term " .. executable)
-  vim.opt_local.relativenumber = false
-  vim.opt_local.number = false
-  vim.cmd(set_bufname)
-  vim.bo.buflisted = false
-  vim.cmd("startinsert")
+  if const.cmake_launch_from_built_binary_directory == true then
```
Comment (Civitasv) [inline lua/cmake-tools/utils.lua:71]: No need to check `== true`, just use `if const.cmake_launch_from_built_binary_directory`

File: lua/cmake-tools/utils.lua [hunk 2]
```diff
@@ -68,16 +68,41 @@ function utils.execute(executable, opts)
...
+          utils.show_cmake_console(opts.cmake_console_position, opts.cmake_console_size)
+          vim.api.nvim_command("cbottom")
+        end
+      end),
+    })
+
+    utils.job:start()
+    return utils.job
+  else
+    local set_bufname = "file " .. opts.bufname
```
Comment (Civitasv) [inline lua/cmake-tools/utils.lua:94]: I think maybe we can delete this chunk, and use cmake_launch_from_built_binary_directory by default.

Comment (rohit-kumar-j) [reply lua/cmake-tools/utils.lua:94]: fully remove the `vim.cmd()` part?

Comment (rohit-kumar-j) [reply lua/cmake-tools/utils.lua:94]: If so, then how about, remove this from the `const` table itself and always launch from built binary directory?

Comment (Civitasv) [reply lua/cmake-tools/utils.lua:94]: Yes, that's what I mean.

Comment (rohit-kumar-j) [reply lua/cmake-tools/utils.lua]: Done.

**Comment by rohit-kumar-j** [comment]
Modified Readme to suit this.

**Comment by Civitasv** [comment]
Forget to say, have you test if we can input for `cin` in the terminal when run the executable?

**Comment by rohit-kumar-j** [comment]
> Forget to say, have you test if we can input for `cin` in the terminal when run the executable?

input for cin? When the program is running?

**Comment by Civitasv** [comment]
Yes. For example:

```cpp
int main()
{
    int a;
    cin << a;
    cout << "A: " << a << endl;
    return 0;
}
```

When I run this program, I should be able to input.

**Comment by rohit-kumar-j** [comment]
Yes, there seems to some [documentation](https://neovim.io/doc/user/job_control.html) on this, although, not in lua. Can you suggest a change for `cin`? 

This does not seem to work:
```lua
function utils.execute(executable, opts)
  -- save all
  vim.cmd("wall")
  -- print("EXECUTABLE", executable)
  utils.job = Job:new({
    command = executable,
    args = {},
    cwd = opts.cmake_launch_path,
    stdin = "pipe",
    on_stdout = vim.schedule_wrap(append_to_cmake_console),
    on_stderr = vim.schedule_wrap(append_to_cmake_console),
    on_exit = vim.schedule_wrap(function(_, code, signal)
      append_to_cmake_console("Exited with code " .. (signal == 0 and code or 128 + signal))
      if code == 0 and signal == 0 then
        if opts.on_success then
          opts.on_success()
        end
      elseif opts.cmake_show_console == "only_on_error" then
        utils.show_cmake_console(opts.cmake_console_position, opts.cmake_console_size)
        vim.api.nvim_command("cbottom")
      end
    end),
  })

```

**Comment by Civitasv** [comment]
Yes, that's just the reason I didn't use it lol.

**Comment by rohit-kumar-j** [comment]
😆 Let me see if there is another method. Like an callback to this

**Comment by Civitasv** [comment]
https://github.com/Civitasv/cmake-tools.nvim/blob/ba1a6cf3f0d6a27ed75c87926153907c1d0de1d6/lua/cmake-tools/utils.lua#L70-L80

I think this should be enough.

**Comment by Civitasv** [comment]
First, we cd to the launch path, then we launch the executable.

**Comment by Civitasv** [comment]
image
And I've tested it use this program, it says Success.

**Comment by rohit-kumar-j** [comment]
> https://github.com/Civitasv/cmake-tools.nvim/blob/ba1a6cf3f0d6a27ed75c87926153907c1d0de1d6/lua/cmake-tools/utils.lua#L70-L80
> 
> I think this should be enough.

image

**Comment by rohit-kumar-j** [comment]
pleanary job control has a [writer](https://github.com/nvim-lua/plenary.nvim/blob/9ac3e9541bbabd9d73663d757e4fe48a675bb054/lua/plenary/job.lua#LL18C56-L18C56) option. 

Seems like it is not exposed in the `:h plenary.nvim-pleanry-job`

**Comment by Civitasv** [comment]
I think the reason for this fault is `vim.cmd(set_bufname)` ?

**Comment by rohit-kumar-j** [comment]
Seems like it.

How about, something like this:  append to buffer if open?

**Comment by Civitasv** [comment]
> How about, something like this: append to buffer if open?

I think that will do, do you know how to do it?

**Comment by rohit-kumar-j** [comment]
Not sure yet. There is a pleanry [writer option](https://github.com/nvim-lua/plenary.nvim/blob/9ac3e9541bbabd9d73663d757e4fe48a675bb054/lua/plenary/job.lua#LL18C56-L18C56) that is not exposed in `:h plenary.nvim-pleanry-job` api. Perhaps this can solve it.

**Comment by Civitasv** [comment]
> Opens mulitple terminals, that are not shown in the bufferline

Why is this happen? I don't quite understand.

**Comment by rohit-kumar-j** [comment]
Seems to be opening multiple buffers as well:
(perhaps behaviour is undefined)

image

**Comment by rohit-kumar-j** [comment]
previously it didn't show them in bufferline.

**Comment by Civitasv** [comment]
Interesting 🧐

**Comment by rohit-kumar-j** [comment]
**writer** seems to work with the api, but it isn't waiting for input(ofcourse I hardcoded it in):

```lua
  utils.job = Job:new({
    command = executable,
    args = {},
    cwd = opts.cmake_launch_path,
    writer = "1",
    on_stdout = vim.schedule_wrap(append_to_cmake_console),
    on_stderr = vim.schedule_wrap(append_to_cmake_console),
    on_exit = vim.schedule_wrap(function(_, code, signal)
      append_to_cmake_console("Exited with code " .. (signal == 0 and code or 118 + signal))
      if code == 0 and signal == 0 then
        if opts.on_success then
          opts.on_success()
        end
      elseif opts.cmake_show_console == "only_on_error" then
        utils.show_cmake_console(opts.cmake_console_position, opts.cmake_console_size)
        vim.api.nvim_command("cbottom")
      end
    end),
  })
```

image

**Comment by rohit-kumar-j** [comment]
This works. Seems like as long as it is a table, it works:
```lua
function temp()
    return {"1"}
end
```
```lua
writer = temp(),
```

**Comment by Civitasv** [comment]
But can it read read from standard input?

**Comment by rohit-kumar-j** [comment]
need to return a stdin to writer in that case. Do you know how to create a function that can return a stdin?

**Comment by Civitasv** [comment]
return io.read?

**Comment by rohit-kumar-j** [comment]
ooh no.. terminal freezes.

**Comment by Civitasv** [comment]
Maybe this api is not supposed to use like that.

**Comment by rohit-kumar-j** [comment]
Yes seems like it does not have full support yet. Will try to do this using native `jobcontrol()` api.

**Comment by rohit-kumar-j** [comment]
@Civitasv, this [vi stackexchange post](https://vi.stackexchange.com/questions/17816/solved-ish-neovim-dont-close-terminal-buffer-after-process-exit) works for not removing output from the terminal buffer.

Using the command `<C-\><C-n>` puts it in inspect mode as read only. Can this be an autocmd when process exits with code xyz?

**Comment by Civitasv** [comment]
Have you seen the effect for pressing <C-\><C-n> on your computer?

**Comment by rohit-kumar-j** [comment]
It shows `--TERMINAL--` in the statusline. No output until in the statusline until full sequence `<C-\><C-n>` and then no output.

**Comment by Civitasv** [comment]
The terminal still disappeared?

**Comment by rohit-kumar-j** [comment]
It does not disappear after `<C-\><C-n>`

However the setbufname is still giving error. Perhaps a pcall(), if something like that exists.

https://github.com/Civitasv/cmake-tools.nvim/assets/37873142/906c5f79-9dae-42dd-a8dc-6801148b35b1

**Comment by Civitasv** [comment]
I wonder if it will be okay if we remove vim.cmd(set_bufname)

**Comment by rohit-kumar-j** [comment]
if we purely do that, then unless we explicitly close the terminal buffer , it does not close the terminals(with `<C-\><C-n>`).
image

**Comment by Civitasv** [comment]
> if we purely do that, then unless we explicitly close the terminal buffer , it does not close the terminals(with `<C-\><C-n>`).
> image

I thinks it should behave like this. And it will disappear if you first enter insert mode and then press any key, am i right?

**Comment by rohit-kumar-j** [comment]
yes

**Comment by rohit-kumar-j** [comment]
Something like this in `init.lua` for persistent output:

```lua
vim.api.nvim_create_autocmd("TermClose", { -- TermClose on proccess end
    group    = vim.api.nvim_create_augroup("cmaketools", { clear = true }),
    callback =  function()
        vim.cmd('call feedkeys("<C-\\><C-n>")')
    end
})
```

**Comment by rohit-kumar-j** [comment]
> Something like this in `init.lua` for persistent output:
>
>```lua
>vim.api.nvim_create_autocmd("TermClose", { -- TermClose on proccess end
>    group    = vim.api.nvim_create_augroup("cmaketools", { clear = true }),
>    callback =  function()
>        vim.cmd('call feedkeys("<C-\\><C-n>")')
>    end
>})
>```

But this does not work: `vim.cmd('call feedkeys("<C-\\><C-n>")')`

**Comment by Civitasv** [comment]
https://github.com/Civitasv/cmake-tools.nvim/blob/ba1a6cf3f0d6a27ed75c87926153907c1d0de1d6/lua/cmake-tools/utils.lua#L80
Actually, I have used this to start insert mode in terminal, maybe we can remove this line too?

**Comment by rohit-kumar-j** [comment]
No, the `startinsert` is useful as it puts in insert mode immediately in insert ready for cin. Even if we remove cin, it still quits.

**Comment by Civitasv** [comment]
> it still quits.

Weird, I cannot produce this situation, it wasn't quit in my computer.

**Comment by rohit-kumar-j** [comment]
I now works! within `init.lua`:
```lua
vim.api.nvim_create_autocmd("TermClose", {
    group    = "cmaketools",
    callback =  function()
        vim.cmd.stopinsert()
        vim.api.nvim_feedkeys('<C-\\><C-n><CR>', 'n', false)
    end
})
```
maybe I can put a `if is_win32 then <create autocommand> end`

**Comment by rohit-kumar-j** [comment]
Or make it optional form the config like `config.experimental = true`

**Comment by Civitasv** [comment]
I think you can use `if is_win32`. No need to add it into configuration I think.

**Comment by rohit-kumar-j** [comment]
How about this, check what bufs are avaliable in the instance and then delete if the buffer exists before spawnning `vim.cmd()`:

```lua
  local all_buffs = vim.api.nvim_list_bufs()
  local temp = " "
  for i, buf_nr in ipairs(all_buffs) do
    local name = vim.api.nvim_buf_get_name(buf_nr)
    local test = string.match(name, opts.cmake_launch_path) == opts.cmake_launch_path
    -- print(test)
    temp = temp .. name .. tostring(test) .. ", "
  end
  print(temp)

  ...
  vim.cmd(prefix ....)
  vim.bo.buflisted = true

  ...
```

It outputs even if it is alive:

First time:

image

Second time:

image

**Comment by Civitasv** [comment]
I think it's feasible.

**Comment by rohit-kumar-j** [comment]
Finally❗ (hope it works as well on linux)

https://github.com/Civitasv/cmake-tools.nvim/assets/37873142/25b31121-9117-48cc-9ec9-366ec197c14b

The buffers seem managed:
image

The output of `:lua vim.pretty_print(vim.api.nvim_list_bufs())` does not increase.

**Comment by rohit-kumar-j** [comment]
I will send a draft PR after a little cleanup

**Comment by Civitasv** [comment]
Good!

**Pull request merged**